File: packages/jetbrains-plugin/src/main/java/com/chainlesschain/ide/PlanReview.java
package com.chainlesschain.ide;

import java.time.Instant;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/** Pure helper for IDE-native plan review editor tabs. */
public final class PlanReview {
    private PlanReview() {}

    public static final int MAX_SNAPSHOT_CHARS = 24000;
    public static final String STATE_SCHEMA = "cc-plan-review/v1";
    public static final int MAX_PERSISTED_STATES = 20;
    public static final int MAX_PERSISTED_ITEMS = 128;

    public static String markdown(
            Map<String, Object> plan,
            String conversationTitle,
            String sessionId,
            Instant updatedAt) {
        Map<String, Object> p = plan != null ? plan : new LinkedHashMap<String, Object>();
        List<Map<String, Object>> items = normalizedItems(p.get("items"));
        StringBuilder sb = new StringBuilder();
        sb.append("# ChainlessChain Plan Review\n\n");
        sb.append("- Conversation: ").append(or(conversationTitle, "Chat")).append("\n");
        sb.append("- Session: ").append(or(sessionId, "(pending)")).append("\n");
        sb.append("- State: ").append(or(p.get("state"), "analyzing")).append("\n");
        sb.append("- Updated: ").append(updatedAt != null ? updatedAt : Instant.now()).append("\n");
        Object risk = p.get("risk");
        if (risk instanceof Map) {
            Map<?, ?> r = (Map<?, ?>) risk;
            sb.append("- Risk: ").append(or(r.get("level"), "unknown"));
            if (r.get("totalScore") != null) sb.append(" (").append(r.get("totalScore")).append(")");
            sb.append("\n");
        }
        if (p.get("note") != null) sb.append("- Note: ").append(p.get("note")).append("\n");
        sb.append("\n## Review Actions\n\n");
        sb.append("Use the Plan card buttons, /approve, /reject, or the chat toolbar actions.\n\n");
        sb.append("## Inline Comments\n\n");
        sb.append("Add comments under the relevant plan item or in Reviewer Notes before choosing an action.\n\n");
        sb.append("## Plan Items\n\n");
        if (items.isEmpty()) {
            sb.append("- (No plan items yet. Blocked write/execute tools will appear here.)\n");
        } else {
            int n = 1;
            for (Map<String, Object> item : items) {
                String tool = string(item.get("tool"));
                sb.append(n++).append(". ");
                if (!tool.isEmpty()) sb.append(tool).append(": ");
                sb.append(or(item.get("title"), "(untitled item)")).append("\n");
                sb.append("   - id: ").append(or(item.get("id"), "")).append("\n");
                sb.append("   - impact: ").append(or(item.get("impact"), "low")).append("\n");
                sb.append("   - status: ").append(or(item.get("status"), "pending")).append("\n");
                sb.append("   - comment:\n");
            }
        }
        sb.append("\n## Reviewer Notes\n\n- ");
        return sb.toString();
    }

    public static Map<String, Object> reviewRecord(
            String action,
            String documentText,
            String conversationId,
            String conversationTitle,
            String sessionId,
            Map<String, Object> plan,
            Instant reviewedAt) {
        Map<String, Object> r = new LinkedHashMap<String, Object>();
        r.put("action", or(action, "review"));
        r.put("reviewedAt", String.valueOf(reviewedAt != null ? reviewedAt : Instant.now()));
        r.put("conversationId", or(conversationId, ""));
        r.put("conversationTitle", or(conversationTitle, ""));
        r.put("sessionId", or(sessionId, ""));
        r.put("planState", plan == null ? "" : or(plan.get("state"), ""));
        r.put("itemCount", normalizedItems(plan == null ? null : plan.get("items")).size());
        r.put("snapshot", trimSnapshot(documentText));
        return r;
    }

    public static Map<String, Object> planEvent(String action, Map<String, Object> review) {
        Map<String, Object> ev = MiniJson.obj();
        ev.put("type", "plan");
        ev.put("action", action);
        if (review != null && review.get("snapshot") instanceof String
                && !String.valueOf(review.get("snapshot")).trim().isEmpty()) {
            ev.put("review", review);
        }
        return ev;
    }

    public static String feedbackPrompt(String action, String documentText) {
        String verb = "regenerate".equals(action)
                ? "Regenerate the plan from scratch"
                : "Revise the plan";
        return verb + " based on this plan review document.\n\n"
                + "Keep plan mode active. Do not execute write or shell tools until "
                + "the revised plan is approved.\n\n"
                + "```markdown\n"
                + trimSnapshot(documentText)
                + "\n```";
    }

    public static String trimSnapshot(String text) {
        String raw = text == null ? "" : text;
        if (raw.length() <= MAX_SNAPSHOT_CHARS) return raw;
        int omitted = raw.length() - MAX_SNAPSHOT_CHARS;
        return raw.substring(0, MAX_SNAPSHOT_CHARS)
                + "\n\n[review snapshot truncated: " + omitted + " chars omitted]";
    }

    /** Build one bounded, versioned plan-review draft/decision snapshot. */
    public static Map<String, Object> persistedState(
            String documentText,
            String conversationId,
            String conversationTitle,
            String sessionId,
            Map<String, Object> plan,
            String status,
            String action,
            Map<String, Object> previous,
            Instant updatedAt) {
        Map<String, Object> prior = normalizePersistedState(previous);
        Map<String, Object> state = new LinkedHashMap<String, Object>();
        state.put("schema", STATE_SCHEMA);
        state.put("revision", prior == null ? 1 : number(prior.get("revision"), 1) + 1);
        state.put("updatedAt", String.valueOf(updatedAt != null ? updatedAt : Instant.now()));
        state.put("conversationId", bounded(conversationId, 256));
        state.put("conversationTitle", bounded(conversationTitle, 256));
        state.put("sessionId", bounded(sessionId, 256));
        state.put("status", or(bounded(status, 64), "draft"));
        state.put("action", bounded(action, 64));
        state.put("plan", sanitizePlanSnapshot(plan));
        state.put("snapshot", trimSnapshot(documentText));
        return normalizePersistedState(state);
    }

    @SuppressWarnings("unchecked")
    public static Map<String, Object> normalizePersistedState(Object value) {
        if (!(value instanceof Map)) return null;
        Map<?, ?> source = (Map<?, ?>) value;
        if (!STATE_SCHEMA.equals(source.get("schema"))) return null;
        Map<String, Object> state = new LinkedHashMap<String, Object>();
        state.put("schema", STATE_SCHEMA);
        state.put("revision", Math.max(1, number(source.get("revision"), 1)));
        state.put("updatedAt", bounded(source.get("updatedAt"), 64));
        state.put("conversationId", bounded(source.get("conversationId"), 256));
        state.put("conversationTitle", bounded(source.get("conversationTitle"), 256));
        state.put("sessionId", bounded(source.get("sessionId"), 256));
        state.put("status", or(bounded(source.get("status"), 64), "draft"));
        state.put("action", bounded(source.get("action"), 64));
        state.put("plan", sanitizePlanSnapshot(
                source.get("plan") instanceof Map
                        ? (Map<String, Object>) source.get("plan") : null));
        state.put("snapshot", trimSnapshot(string(source.get("snapshot"))));
        return persistedIdentity(state).isEmpty() ? null : state;
    }

    public static Map<String, Object> findPersistedState(
            Object values, String sessionId, String conversationId) {
        String wanted = persistedIdentity(sessionId, conversationId);
        if (wanted.isEmpty() || !(values instanceof List)) return null;
        List<?> list = (List<?>) values;
        for (int i = list.size() - 1; i >= 0; i--) {
            Map<String, Object> state = normalizePersistedState(list.get(i));
            if (state != null && wanted.equals(persistedIdentity(state))) return state;
        }
        return null;
    }

    public static List<Map<String, Object>> upsertPersistedState(
            Object values, Map<String, Object> value) {
        Map<String, Object> state = normalizePersistedState(value);
        List<Map<String, Object>> kept = new ArrayList<Map<String, Object>>();
        String identity = state == null ? "" : persistedIdentity(state);
        if (values instanceof List) {
            for (Object raw : (List<?>) values) {
                Map<String, Object> entry = normalizePersistedState(raw);
                if (entry != null && !identity.equals(persistedIdentity(entry))) {
                    kept.add(entry);
                }
            }
        }
        if (state != null) kept.add(state);
        if (kept.size() <= MAX_PERSISTED_STATES) return kept;
        return new ArrayList<Map<String, Object>>(
                kept.subList(kept.size() - MAX_PERSISTED_STATES, kept.size()));
    }

    public static Map<String, Object> sanitizePlanSnapshot(Map<String, Object> plan) {
        Map<String, Object> source = plan != null ? plan : new LinkedHashMap<String, Object>();
        Map<String, Object> snapshot = new LinkedHashMap<String, Object>();
        snapshot.put("active", Boolean.TRUE.equals(source.get("active")));
        snapshot.put("state", bounded(source.get("state"), 128));
        List<Map<String, Object>> items = normalizedItems(source.get("items"));
        List<Map<String, Object>> boundedItems = new ArrayList<Map<String, Object>>();
        for (int i = 0; i < items.size() && i < MAX_PERSISTED_ITEMS; i++) {
            Map<String, Object> item = items.get(i);
            Map<String, Object> row = new LinkedHashMap<String, Object>();
            row.put("id", bounded(item.get("id"), 160));
            row.put("title", bounded(item.get("title"), 512));
            row.put("tool", bounded(item.get("tool"), 160));
            row.put("impact", bounded(item.get("impact"), 64));
            row.put("status", bounded(item.get("status"), 64));
            boundedItems.add(row);
        }
        snapshot.put("items", boundedItems);
        if (source.get("note") != null) snapshot.put("note", bounded(source.get("note"), 2000));
        if (source.get("risk") instanceof Map) {
            Map<?, ?> inputRisk = (Map<?, ?>) source.get("risk");
            Map<String, Object> risk = new LinkedHashMap<String, Object>();
            risk.put("level", bounded(inputRisk.get("level"), 64));
            if (inputRisk.get("totalScore") instanceof Number) {
                risk.put("totalScore", inputRisk.get("totalScore"));
            }
            snapshot.put("risk", risk);
        }
        return snapshot;
    }

    @SuppressWarnings("unchecked")
    public static List<Map<String, Object>> normalizedItems(Object value) {
        List<Map<String, Object>> out = new ArrayList<Map<String, Object>>();
        if (!(value instanceof List)) return out;
        int n = 1;
        for (Object item : (List<?>) value) {
            Map<String, Object> row = new LinkedHashMap<String, Object>();
            if (item instanceof Map) {
                Map<?, ?> m = (Map<?, ?>) item;
                row.put("id", or(m.get("id"), "item-" + n));
                row.put("title", or(firstPresent(m, "title", "text"), "(untitled item)"));
                row.put("tool", or(m.get("tool"), ""));
                row.put("impact", or(firstPresent(m, "impact", "estimatedImpact"), "low"));
                row.put("status", or(m.get("status"), "pending"));
            } else {
                row.put("id", "item-" + n);
                row.put("title", String.valueOf(item));
                row.put("tool", "");
                row.put("impact", "low");
                row.put("status", "pending");
            }
            out.add(row);
            n++;
        }
        return out;
    }

    private static Object firstPresent(Map<?, ?> m, String... keys) {
        for (String key : keys) {
            Object v = m.get(key);
            if (v != null && !String.valueOf(v).isEmpty()) return v;
        }
        return null;
    }

    private static String persistedIdentity(Map<String, Object> state) {
        return persistedIdentity(string(state.get("sessionId")), string(state.get("conversationId")));
    }

    private static String persistedIdentity(String sessionId, String conversationId) {
        String session = bounded(sessionId, 256);
        if (!session.isEmpty()) return "session:" + session;
        String conversation = bounded(conversationId, 256);
        return conversation.isEmpty() ? "" : "conversation:" + conversation;
    }

    private static int number(Object value, int fallback) {
        return value instanceof Number ? ((Number) value).intValue() : fallback;
    }

    private static String bounded(Object value, int limit) {
        String s = string(value);
        return s.length() <= limit ? s : s.substring(0, limit);
    }

    private static String or(Object value, String fallback) {
        String s = string(value);
        return s.isEmpty() ? fallback : s;
    }

    private static String string(Object value) {
        return value == null ? "" : String.valueOf(value);
    }
}

File: packages/vscode-extension/src/chat/plan-review.js
const MAX_REVIEW_SNAPSHOT_CHARS = 24000;
const PLAN_REVIEW_STATE_SCHEMA = "cc-plan-review/v1";
const MAX_PERSISTED_PLAN_REVIEWS = 20;
const MAX_PERSISTED_PLAN_ITEMS = 128;
const MAX_PLAN_REVIEW_COMMENTS = 64;
const MAX_PLAN_REVIEW_COMMENT_CHARS = 2000;

function normalizePlanItems(items) {
  return (Array.isArray(items) ? items : []).map((item, index) => ({
    id: stringOr(item?.id, `item-${index + 1}`),
    title: stringOr(item?.title, "(untitled item)"),
    tool: stringOr(item?.tool, ""),
    impact: stringOr(item?.impact, "low"),
    status: stringOr(item?.status, "pending"),
    turn:
      Number.isInteger(item?.turn) && item.turn > 0 ? Number(item.turn) : null,
    toolUseId: stringOr(item?.toolUseId || item?.tool_use_id, ""),
    startedAt: stringOr(item?.startedAt || item?.started_at, ""),
    completedAt: stringOr(item?.completedAt || item?.completed_at, ""),
    error: stringOr(item?.error, ""),
  }));
}

function stringOr(value, fallback = "") {
  if (value == null) return fallback;
  const s = String(value);
  return s || fallback;
}

function formatDate(when = new Date()) {
  try {
    return when instanceof Date
      ? when.toISOString()
      : new Date(when).toISOString();
  } catch {
    return new Date().toISOString();
  }
}

function formatPlanReviewMarkdown(plan = {}, opts = {}) {
  const items = normalizePlanItems(plan.items);
  const risk = plan.risk && typeof plan.risk === "object" ? plan.risk : null;
  const title = stringOr(opts.conversationTitle, "Chat");
  const sessionId = stringOr(opts.sessionId, "(pending)");
  const updatedAt = formatDate(opts.updatedAt);
  const lines = [
    "# ChainlessChain Plan Review",
    "",
    `- Conversation: ${title}`,
    `- Session: ${sessionId}`,
    `- State: ${stringOr(plan.state, "analyzing")}`,
    `- Updated: ${updatedAt}`,
  ];
  if (risk) {
    lines.push(
      `- Risk: ${stringOr(risk.level, "unknown")}` +
        (risk.totalScore == null ? "" : ` (${risk.totalScore})`),
    );
  }
  if (plan.note) lines.push(`- Note: ${String(plan.note)}`);
  lines.push(
    "",
    "## Review Actions",
    "",
    "Use the editor commands:",
    "",
    "- ChainlessChain: Approve Plan Review",
    "- ChainlessChain: Request Plan Changes",
    "- ChainlessChain: Regenerate Plan",
    "- ChainlessChain: Reject Plan Review",
    "",
    "## Inline Comments",
    "",
    "Add comments under the relevant plan item or in Reviewer Notes before choosing an action.",
    "",
    "## Plan Items",
    "",
  );
  if (!items.length) {
    lines.push(
      "- (No plan items yet. Blocked write/execute tools will appear here.)",
    );
  } else {
    items.forEach((item, index) => {
      const tool = item.tool ? `${item.tool}: ` : "";
      lines.push(
        `${index + 1}. ${tool}${item.title}`,
        `   - id: ${item.id}`,
        `   - impact: ${item.impact}`,
        `   - status: ${item.status}`,
      );
      const progress = formatPlanItemProgress(item);
      if (progress) lines.push(`   - progress: ${progress}`);
      if (item.error) lines.push(`   - error: ${item.error}`);
      lines.push("   - comment:");
    });
  }
  lines.push("", "## Reviewer Notes", "", "- ");
  return lines.join("\n");
}

function trimReviewSnapshot(text, limit = MAX_REVIEW_SNAPSHOT_CHARS) {
  const raw = String(text || "");
  if (raw.length <= limit) return raw;
  const omitted = raw.length - limit;
  return (
    raw.slice(0, limit) +
    `\n\n[review snapshot truncated: ${omitted} chars omitted]`
  );
}

function buildPlanReviewRecord({
  action,
  documentText,
  conversationId,
  conversationTitle,
  sessionId,
  plan,
  revision,
  turn,
  now = new Date(),
} = {}) {
  const comments = extractPlanReviewComments(documentText, {
    turn: turn || latestPlanTurn(plan),
  });
  const record = {
    action: stringOr(action, "review"),
    reviewedAt: formatDate(now),
    conversationId: stringOr(conversationId, ""),
    conversationTitle: stringOr(conversationTitle, ""),
    sessionId: stringOr(sessionId, ""),
    planState: stringOr(plan?.state, ""),
    itemCount: normalizePlanItems(plan?.items).length,
    snapshot: trimReviewSnapshot(documentText),
    comments,
  };
  if (Number.isInteger(revision) && revision > 0) {
    record.revision = revision;
  }
  if (Number.isInteger(turn) && turn > 0) record.turn = turn;
  return record;
}

function latestPlanTurn(plan) {
  return normalizePlanItems(plan?.items).reduce(
    (latest, item) => Math.max(latest, item.turn || 0),
    0,
  );
}

function formatPlanItemProgress(item) {
  const parts = [];
  if (item.turn) parts.push(`turn ${item.turn}`);
  if (item.toolUseId) parts.push(`tool use ${item.toolUseId}`);
  if (item.startedAt) parts.push(`started ${item.startedAt}`);
  if (item.completedAt) parts.push(`completed ${item.completedAt}`);
  return parts.join("; ");
}

function fileReference(text) {
  const match = String(text || "").match(
    /((?:[A-Za-z]:[\\/])?(?:[\w@.+()-]+[\\/])*[\w@.+()-]+\.[A-Za-z0-9]+):(\d+)(?::(\d+))?/,
  );
  if (!match) return {};
  return {
    file: match[1].slice(0, 1024),
    line: Number(match[2]),
    ...(match[3] ? { column: Number(match[3]) } : {}),
  };
}

function normalizePlanReviewComment(value, index = 0) {
  if (!value || typeof value !== "object") return null;
  const text = boundedString(value.text, MAX_PLAN_REVIEW_COMMENT_CHARS).trim();
  if (!text) return null;
  const ref = fileReference(text);
  const positive = (candidate) => {
    const number = Number(candidate);
    return Number.isInteger(number) && number > 0 ? number : null;
  };
  return {
    id: boundedString(value.id, 160) || `comment-${index + 1}`,
    sourceLine: positive(value.sourceLine),
    itemId: boundedString(value.itemId, 160) || null,
    text,
    file: boundedString(value.file, 1024) || ref.file || null,
    line: positive(value.line) || ref.line || null,
    column: positive(value.column) || ref.column || null,
    turn: positive(value.turn),
  };
}

function normalizePlanReviewComments(values) {
  return (Array.isArray(values) ? values : [])
    .slice(0, MAX_PLAN_REVIEW_COMMENTS)
    .map(normalizePlanReviewComment)
    .filter(Boolean);
}

function extractPlanReviewComments(documentText, { turn = null } = {}) {
  const lines = String(documentText || "").split(/\r?\n/);
  const comments = [];
  let itemId = null;
  let reviewerNotes = false;
  const push = (text, sourceLine, targetItemId) => {
    if (comments.length >= MAX_PLAN_REVIEW_COMMENTS) return;
    const comment = normalizePlanReviewComment(
      { text, sourceLine, itemId: targetItemId, turn },
      comments.length,
    );
    if (comment) comments.push(comment);
  };

  for (let index = 0; index < lines.length; index += 1) {
    const line = lines[index];
    const heading = line.match(/^##\s+(.+?)\s*$/);
    if (heading) {
      reviewerNotes = /^Reviewer Notes$/i.test(heading[1]);
      itemId = null;
      continue;
    }
    if (/^\d+\.\s+/.test(line)) {
      reviewerNotes = false;
      itemId = null;
      continue;
    }
    const idMatch = line.match(/^\s*-\s*id:\s*(.+?)\s*$/i);
    if (idMatch) {
      itemId = boundedString(idMatch[1], 160);
      continue;
    }
    const commentMatch = line.match(/^\s*-\s*comment:\s*(.*)$/i);
    if (commentMatch) {
      const parts = [commentMatch[1].trim()].filter(Boolean);
      let cursor = index + 1;
      while (cursor < lines.length) {
        const continuation = lines[cursor];
        if (
          /^\s*-\s*(?:id|impact|status|progress|error|comment):/i.test(
            continuation,
          ) ||
          /^\d+\.\s+/.test(continuation) ||
          /^##\s+/.test(continuation)
        ) {
          break;
        }
        if (continuation.trim()) parts.push(continuation.trim());
        cursor += 1;
      }
      push(parts.join("\n"), index + 1, itemId);
      index = cursor - 1;
      continue;
    }
    if (reviewerNotes) {
      const note = line.match(/^\s*[-*]\s+(.+?)\s*$/);
      if (note) push(note[1], index + 1, null);
    }
  }
  return comments;
}

function mergePlanReviewProgress(documentText, plan = {}) {
  const items = new Map(
    normalizePlanItems(plan.items).map((item) => [item.id, item]),
  );
  const lines = String(documentText || "").split(/\r?\n/);
  const out = [];
  let itemId = null;
  for (let index = 0; index < lines.length; index += 1) {
    const line = lines[index];
    const idMatch = line.match(/^\s*-\s*id:\s*(.+?)\s*$/i);
    if (idMatch) itemId = idMatch[1];
    if (/^\d+\.\s+/.test(line) || /^##\s+/.test(line)) itemId = null;
    const statusMatch = line.match(/^(\s*-\s*status:)\s*.*$/i);
    const item = itemId ? items.get(itemId) : null;
    if (!statusMatch || !item) {
      if (item && /^\s*-\s*(?:progress|error):/i.test(line)) {
        continue;
      }
      out.push(line);
      continue;
    }
    out.push(`${statusMatch[1]} ${item.status}`);
    const progress = formatPlanItemProgress(item);
    const indent = statusMatch[1].match(/^\s*/)?.[0] || "";
    if (progress) out.push(`${indent}- progress: ${progress}`);
    if (item.error) out.push(`${indent}- error: ${item.error}`);
    while (
      index + 1 < lines.length &&
      /^\s*-\s*(?:progress|error):/i.test(lines[index + 1])
    ) {
      index += 1;
    }
  }
  return out.join("\n");
}

function buildPlanReviewFeedbackPrompt(action, documentText) {
  const verb =
    action === "regenerate"
      ? "Regenerate the plan from scratch"
      : "Revise the plan";
  return [
    `${verb} based on this plan review document.`,
    "",
    "Keep plan mode active. Do not execute write or shell tools until the revised plan is approved.",
    "",
    "```markdown",
    trimReviewSnapshot(documentText),
    "```",
  ].join("\n");
}

function boundedString(value, limit = 512) {
  return stringOr(value, "").slice(0, limit);
}

function sanitizePlanSnapshot(plan = {}) {
  const source = plan && typeof plan === "object" ? plan : {};
  const snapshot = {
    active: source.active === true,
    state: boundedString(source.state, 128),
    items: normalizePlanItems(source.items)
      .slice(0, MAX_PERSISTED_PLAN_ITEMS)
      .map((item) => ({
        id: boundedString(item.id, 160),
        title: boundedString(item.title, 512),
        tool: boundedString(item.tool, 160),
        impact: boundedString(item.impact, 64),
        status: boundedString(item.status, 64),
        ...(item.turn ? { turn: item.turn } : {}),
        ...(item.toolUseId
          ? { tool_use_id: boundedString(item.toolUseId, 160) }
          : {}),
        ...(item.startedAt
          ? { started_at: boundedString(item.startedAt, 64) }
          : {}),
        ...(item.completedAt
          ? { completed_at: boundedString(item.completedAt, 64) }
          : {}),
        ...(item.error ? { error: boundedString(item.error, 1000) } : {}),
      })),
  };
  if (source.note != null) snapshot.note = boundedString(source.note, 2000);
  if (source.risk && typeof source.risk === "object") {
    snapshot.risk = {
      level: boundedString(source.risk.level, 64),
      ...(Number.isFinite(Number(source.risk.totalScore))
        ? { totalScore: Number(source.risk.totalScore) }
        : {}),
    };
  }
  return snapshot;
}

function persistedPlanReviewIdentity(value = {}) {
  const sessionId = boundedString(value.sessionId, 256);
  if (sessionId) return `session:${sessionId}`;
  const conversationId = boundedString(value.conversationId, 256);
  return conversationId ? `conversation:${conversationId}` : "";
}

function normalizePersistedPlanReview(value) {
  if (
    !value ||
    typeof value !== "object" ||
    value.schema !== PLAN_REVIEW_STATE_SCHEMA
  ) {
    return null;
  }
  const normalized = {
    schema: PLAN_REVIEW_STATE_SCHEMA,
    revision:
      Number.isInteger(value.revision) && value.revision > 0
        ? value.revision
        : 1,
    updatedAt: formatDate(value.updatedAt),
    conversationId: boundedString(value.conversationId, 256),
    conversationTitle: boundedString(value.conversationTitle, 256),
    sessionId: boundedString(value.sessionId, 256),
    status: boundedString(value.status, 64) || "draft",
    action: boundedString(value.action, 64),
    plan: sanitizePlanSnapshot(value.plan),
    snapshot: trimReviewSnapshot(value.snapshot),
  };
  normalized.comments = normalizePlanReviewComments(
    Array.isArray(value.comments)
      ? value.comments
      : extractPlanReviewComments(normalized.snapshot, {
          turn: latestPlanTurn(normalized.plan),
        }),
  );
  return persistedPlanReviewIdentity(normalized) ? normalized : null;
}

function buildPersistedPlanReview({
  documentText,
  conversationId,
  conversationTitle,
  sessionId,
  plan,
  status = "draft",
  action = "",
  previous,
  comments,
  turn,
  now = new Date(),
} = {}) {
  const prior = normalizePersistedPlanReview(previous);
  return normalizePersistedPlanReview({
    schema: PLAN_REVIEW_STATE_SCHEMA,
    revision: prior ? prior.revision + 1 : 1,
    updatedAt: formatDate(now),
    conversationId,
    conversationTitle,
    sessionId,
    status,
    action,
    plan,
    snapshot: documentText,
    comments:
      comments ||
      extractPlanReviewComments(documentText, {
        turn: turn || latestPlanTurn(plan),
      }),
  });
}

function findPersistedPlanReview(list, identity = {}) {
  const wanted = persistedPlanReviewIdentity(identity);
  if (!wanted) return null;
  const values = Array.isArray(list) ? list : [];
  for (let i = values.length - 1; i >= 0; i -= 1) {
    const state = normalizePersistedPlanReview(values[i]);
    if (state && persistedPlanReviewIdentity(state) === wanted) return state;
  }
  return null;
}

function upsertPersistedPlanReview(
  list,
  value,
  limit = MAX_PERSISTED_PLAN_REVIEWS,
) {
  const state = normalizePersistedPlanReview(value);
  if (!state) return (Array.isArray(list) ? list : []).slice(-limit);
  const identity = persistedPlanReviewIdentity(state);
  const kept = (Array.isArray(list) ? list : [])
    .map(normalizePersistedPlanReview)
    .filter(
      (entry) => entry && persistedPlanReviewIdentity(entry) !== identity,
    );
  kept.push(state);
  return kept.slice(-Math.max(1, Number(limit) || MAX_PERSISTED_PLAN_REVIEWS));
}

module.exports = {
  MAX_REVIEW_SNAPSHOT_CHARS,
  MAX_PERSISTED_PLAN_REVIEWS,
  MAX_PLAN_REVIEW_COMMENTS,
  PLAN_REVIEW_STATE_SCHEMA,
  buildPlanReviewFeedbackPrompt,
  buildPlanReviewRecord,
  buildPersistedPlanReview,
  findPersistedPlanReview,
  extractPlanReviewComments,
  formatPlanReviewMarkdown,
  mergePlanReviewProgress,
  normalizePlanReviewComments,
  normalizePersistedPlanReview,
  normalizePlanItems,
  sanitizePlanSnapshot,
  trimReviewSnapshot,
  upsertPersistedPlanReview,
};
